examples/examples/data-channels-flow-control/data-channels-flow-control.rs

```diff
@@ -1,368 +1,263 @@
         .build();
 
-    // Prepare the configuration
+    let ice_servers = vec![RTCIceServer {
+        ..Default::default()
+    }];
+
     let config = RTCConfiguration {
-        ice_servers: vec![RTCIceServer {
-            //urls: vec!["stun:stun.l.google.com:19302".to_owned()],
-            ..Default::default()
-        }],
+        ice_servers,
         ..Default::default()
     };
 
-    // Create a new RTCPeerConnection
-    let pc = Arc::new(api.new_peer_connection(config).await?);
+    api.new_peer_connection(config).await.unwrap()
+}
 
-    let ordered = Some(false);
-    let max_retransmits = Some(0u16);
+async fn create_requester() -> RTCPeerConnection {
+    let pc = create_peer_connection().await;
 
     let options = Some(RTCDataChannelInit {
-        ordered,
-        max_retransmits,
+        ordered: Some(false),
+        max_retransmits: Some(0u16),
         ..Default::default()
     });
 
-    let (send_more_ch_tx, mut send_more_ch_rx) = tokio::sync::mpsc::channel::<()>(1);
-    let send_more_ch_tx = Arc::new(send_more_ch_tx);
+    let (more_can_be_sent, mut maybe_more_can_be_sent) = tokio::sync::mpsc::channel(1);
+    let dc = pc.create_data_channel("data", options).await.unwrap();
+
+    let shared_dc = dc.clone();
+    dc.on_open(Box::new(|| {
+        Box::pin(async move {
+            println!("requester :: on_open");
 
-    // Create a datachannel with label 'data'
-    let dc = pc.create_data_channel("data", options).await?;
+            let task = tokio::spawn(async move {
+                let buf = Bytes::from_static(&[0u8; 1024]);
 
-    // Register channel opening handling
+                loop {
+                    if shared_dc.send(&buf).await.is_err() {
+                        break;
+                    }
 
-    // no need to downgrade this to Weak, since on_open is FnOnce callback
-    let dc2 = Arc::clone(&dc);
-    dc.on_open(Box::new(|| {
-        println!(
-            "OnOpen-{}-{} : Start sending a series of 1024-byte packets as fast as it can",
-            dc2.label(),
-            dc2.id()
-        );
-
-        tokio::spawn(async move {
-            let buf = Bytes::from_static(&[0u8; 1024]);
-            loop {
-                if dc2.send(&buf).await.is_err() {
-                    break;
-                }
+                    let buffered_amount = shared_dc.buffered_amount().await;
 
-                tokio::time::sleep(Duration::from_micros(1)).await;
-                let buffered_amount = dc2.buffered_amount().await;
-                if buffered_amount + buf.len() > MAX_BUFFERED_AMOUNT {
-                    let _ = send_more_ch_rx.recv().await;
+                    if buffered_amount + buf.len() > MAX_BUFFERED_AMOUNT {
+                        let _ = maybe_more_can_be_sent.recv().await;
+                    }
                 }
-            }
-            println!("exit on_open");
-        });
-        Box::pin(async {})
+            });
+
+            task.await.unwrap();
```

Comment:
This is not good, we don't want to block the code that calls that awaits this future on this tasks. The future returned from `on_open` must complete fairly quickly since it awaited within a core part of the code. Blocking for a long time is liable to break stuff

Reply:
It is written somewhere in the documentation that this callback should return future, that "must complete fairly quickly since it awaited within a core part of the code"? Maybe we should make this callback not async, to avoid unexpected blocking?